### PR TITLE
cli: manage seed migrations as sql files (close #2431)

### LINKED ISSUE #2431: managing seed or data-only migrations
Often one would like to add some data into tables as part of the db init process. We should document how to achieve that.

- head to run SQL
- add the insert sql statement
- hit ~track this~ This is a migration
- Run will insert data and create a migration file

**Edit**: 
The above task is done. There are further suggestions for managing seed data migrations in the comments. Leaving this issue open for that discussion

### PR DESCRIPTION
**Pull request by scriptonist**
Please preview it [here](https://github.com/scriptonist/graphql-engine/blob/scriptonist-rfc-seed-data/rfcs/seed-data.md)

Big shoutout to @GavinRay97 for kicking of the discussions with https://github.com/hasura/graphql-engine/pull/3614

### CODE REVIEW

**Comment by rikinsk** [comment]
I agree with Shahidh. Having a seed command that manages a "seed" directory sounds good enough.  It doesnt introduce any new concepts as it seems to be a natural extenstion to the migrations concepts.

**Comment by scriptonist** [comment]
I think most of us think approach 1 is the way to move ahead. Now, I would like to put the following in perspective. 

What kind of data is the seed command supposed to operate on?

Let me try to elaborate on this, say I'm building an app and I want to seed some data. Then I can use the seed command for populating the seeds, note that this data is application-specific.

Also, From what I understand seed migrations is completely different from schema migrations and are not treated the same (from rails point of view). I also have a similar feeling that in the case of seed, we don't need the sophistication that comes with schema migrations in the form of versioning and other things.

What I'm puzzled about is, should the seed API take into account anything outside the user's applications scope? For example, If Hasura wants to populate some critical data then should seed API account for managing that as well? The reason why I'm concerned about this is since it is critical data we are dealing with, we might want to treat it similar to schema migrations.

In a nutshell, what I'm trying to ask is that, 
should the seed command be introduced exclusively to operate on application data or otherwise?

**Comment by rikinsk** [comment]
> in the case of seed, we don't need the sophistication that comes with schema migrations in the form of versioning and other things.

I agree. The scope of seed command should apply all avaialble seed files. I cant think of any typical cases where I would need seed data to be versioned. If we ever are presented with such a situation, we can think about it then.

> If Hasura wants to populate some critical data then should seed API account for managing that as well?

The seed data should only be concerned with the application data. Any other data critical for Hasura should ideally be managed via metadata or should otherwise be part of the migrations workflow

**Comment by coco98** [comment]
@scriptonist @rikinsk Can we add `hasura console` support for seed data migrations in this as well?

**Comment by rikinsk** [comment]
@coco98 Sure. But I ddidnt want to block this. Console support can be added separately when we have someone who can take it up. Unless you want them to go out together only and are ok with like a couple weeks delay

**Comment by shahidhk** [comment]
@scriptonist Let's go ahead with approach 1, will you update the RFC with proposed command structure and example usages before getting started?

**Comment by scriptonist** [comment]
Sure @shahidhk, Since the RFC was lacking an implementation, I went ahead and did it to get a feeling of how things would work and get a better understanding.

The result is a new command called `seed` and two subcommands `create` and `apply`. 

- `hasura seed create <name>` will create a new seed file and open it in the default editor.
-  `hasura seed apply` will try to apply all migrations files in the `seeds/` directory.

In addition to that, a new API which can probably be used by the console to create seed files was added too.

This is an initial implementation, which is not complete and may have a handful of limitations. We can improve and better shape it with respect to what folks think.

**Comment by shahidhk** [comment]
@arvi3411301 Review this too.

File: cli/api/seed.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	Filename string `json:"filename,omitempty"`
+}
+
+// CreateSeedResponse --
+type CreateSeedResponse struct {
+	Message string `json:"message,omitempty"`
+}
+
+// CreateSeedAPIHandler --
+func CreateSeedAPIHandler(c *gin.Context) {
```
[the remark this answers is not in the document]
Comment (scriptonist) [reply cli/api/seed.go]: Should we do this?  I'm thinking we can have another API which can apply seeds 🤔 and let this just handle the creation of seed files. Thoughts?

**Comment by GavinRay97** [comment]
I want to leave a comment about versioned seeds.

The reasoning behind versioned seeds is basically this:

Let's say you write some seed data, where you explicitly insert DB rows with manual ID's so that you can also insert associations by referencing the parent IDs.

When you add more seed data later, if you do not have a way to differentiate between seed data that has already been applied, it will attempt to re-insert the previously added seeds and violate duplicate primary key constraints.

Typically when you seed, you want to make sure that each seed file is applied once and only once, and that the seed command is idempotent. Idempotency is really important when designing applications and systems.
